# Homework 3

* Assigned: 03/07 Thursday

* Due: 04/04 Thursday, 10:00 AM via. Instabase + Gradescope

* Value: 3.75% of your grade

In this assignment it's time to get real! You'll first flex your SQL muscles and perform analyses similar to HW2's by writing SQL and reflecting on the experience. You will then perform some normalization.

## Announcement

Homework 3 has two parts. __Part 1__ requires you to __complete and submit via Instabase__ by the token provided. We have setup the environment for you. Part 2 is a written task requiring a hard copy to be submitted on Gradescope.

__Please do not revise any existing block or variable without instruction.__ It may cause failure for running grading script, which will result in some grade deduction. Query quality and correctness will also be part of the grading rubric. We will grade mannually to allow some partial credits.

## Token Setup

In [29]:
# Your columbia uni that is used in SSOL
#
# IMPORTANT:  make sure this is consistent with the uni/alias used as your @columbia.edu email in SSOL
#
UNI = "zd1111"

# your instabase username (if you go to the instabase homepage, your username should be in the URL)
USER = "zd2221"

# your repository name containing 
REPO = "myrepo"


## Database Connection

Run following command to connect to database. You should see "Connected to ..." for successful connection.

In [30]:
# connect to db in public repo
#ib.connect_db('ib://ewu/w4111-public/databases/w4111')

ib.connect_db("postgresql://student:w4111student@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111")

Connected to: postgresql://student:w4111student@w4111.cisxo09blonu.us-east-1.rds.amazonaws.com/w4111


We are using a 1mil-tuple sample** of the iowa liquor sales data for this assignment, with column names and datatype (in format of [name] [datatype]) as:

  - address char(256),
  - bottle_volume_ml integer,
  - category char(256),
  - category_name char(256),
  - city char(256),
  - county char(256),
  - county_number char(256),
  - date date,
  - im_desc char(256),
  - invoice_line_no char(256),
  - itemno integer,
  - name char(256),
  - pack integer,
  - sale_bottles integer,
  - sale_dollars double precision,
  - sale_gallons double precision,
  - sale_liters double precision,
  - state_bottle_cost double precision,
  - state_bottle_retail double precision,
  - store integer,
  - store_location char(256),
  - store_location_address char(256),
  - store_location_city char(256),
  - store_location_zip char(256),
  - vendor_name char(256),
  - vendor_no integer,
  - zipcode text
  
A description of column names can be found here: https://data.iowa.gov/Economy/Iowa-Liquor-Sales/m3tr-qhgy.

You may run this query to have some ideas on the schema you are dealing with:

In [5]:
%%sql
select * from iowa limit 1;

1 rows affected.


address  bottle_volume_ml  \
0  228 N MAIN ST                                 ...               600   

   category                                      category_name  \
0   1082900  MISC. IMPORTED CORDIALS & LIQUEURS            ...   

                                                city  \
0  MONTICELLO                                    ...   

                                              county  county_number  \
0  Jones                                         ...             53   

         date                                            im_desc  \
0  2013-03-19  Hpnotiq Mini                                  ...   

                                     invoice_line_no   ...     sale_liters  \
0  S11179000142                                  ...   ...             0.6   

  state_bottle_cost  state_bottle_retail  store  \
0             12.44                18.66   3495   

                              store_location_address  \
0  228 N MAIN ST                                 ...   

                                 store_location_city  \
0  MONTICELLO                                    ...   

                                  store_location_zip  \
0  52310                                         ...   

                                         vendor_name  vendor_no  zipcode  
0  Heaven Hill Brands                            ...        259    52310  

[1 rows x 26 columns]

__Disclaimer: this course does not condone drinking, we are using this dataset because it is a common format for a sales transaction log in a silghtly more accessible domain than typical bank transactions__.

## Part I: SQL, the sequel

### Jupyter Notes: _Read these carefully_

* You **may** create new IPython notebook cells to use for e.g. testing, debugging, exploring, etc.- this is encouraged in fact!- **just make sure that you run the final cell to submit your results**
  * you can press shift+enter to execute to code in the cell that your cursor is in.
* When you see `In [*]:` to the left of the cell you are executing, this means that the code / query is _running_. Please wait for the execution to complete
    * **If the cell is hanging- i.e. running for too long: you can restart the kernel**
    * To restart kernel using the menu bar: "Kernel >> Restart >> Clear all outputs & restart"), then re-execute cells from the top
* _Have fun!_

__Now, please write the SQL query for each of the following questions.__

__Fill answers according to instruction in each question. Make sure to submit your query also (instruction given).__

Note: Some queries will take a few minutes to run on your virtual machine. If your query is running for more than ~10 minutes, you either did something wrong, 
or you may need to create tables with your temp tables to run the query more efficiently. 

For example: If you use the results of a sub-query multiple times, it 
can help to use SELECT ... INTO ... then reference that table.


### __(2 points) Q1.1: Which store had the second most sales in terms of total records in the table?__
__You will receive ZERO if you use offset, but you could use offset to verify your answers. Notice that below are some follow ups for this question, so make sure you got the correct answer.__

In [9]:
%%sql
-- Complete your query here
select store, count(store) as count 
from iowa
group by store
order by count desc
OFFSET 1 ROWS
FETCH NEXT 1 ROWS only


1 rows affected.


store  count
0   2190   7802

In [31]:
# Please replace None with the result store and record number here
# submit query in string format
q1 = {
    'store': 2190,
    'count': 7802,
    'query': """select store, count(store) as count 
from iowa
group by store
order by count desc
OFFSET 1 ROWS
FETCH NEXT 1 ROWS only"""
}

### __(2 points) Q1.2: At the store with the second most total records (you may use the answer for Q1.1), what was the vendor number with most sale records? (The vendor that has the most records in the table?)__

In [ ]:
%%sql
-- Complete your query here
with temp1 as (select vendor_no, count(vendor_no) as count from iowa where store=2190 group by vendor_no)
select vendor_no, count from temp1 where count=(select max(count) from temp1)

In [32]:
# please replace None with the result vendor_no and according sum here
# submit query in string format
q2 = {
    'vendor_no': 260,
    'count': 1393,
    'query': """with temp1 as (select vendor_no, count(vendor_no) as count from iowa where store=2190 group by vendor_no) select vendor_no, count from temp1 where count=(select max(count) from temp1)"""
}

### __(2 points) Q1.3: For the store 2130, 2152, 2190 and 2191, write a query to find the most purchased sale_bottles one time (for each row, we regard it as one time). __

__The order of returned tuples (store, sale_bottles) should be from 2130 to 2191. You will receive ZERO if you use group by or any aggregations.__

In [23]:
%%sql
-- your query here
SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2130
ORDER BY sale_bottles desc
LIMIT 1
) as temp1

union 

SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2152
ORDER BY sale_bottles desc
LIMIT 1
) as temp2

union 

SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2190
ORDER BY sale_bottles desc
LIMIT 1
) as temp3

union 

SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2191
ORDER BY sale_bottles desc
LIMIT 1
) as temp4



4 rows affected.


store  sale_bottles
0   2190           480
1   2191           540
2   2152            36
3   2130           600

In [33]:
# please replace None with the result accordingly here
# submit query in string format
q3 = {
    1: {'store': 2130, 'sale_bottles': 600},
    2: {'store': 2152, 'sale_bottles': 36},
    3: {'store': 2190, 'sale_bottles': 480},
    4: {'store': 2191, 'sale_bottles': 540},
    'query': """SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2130
ORDER BY sale_bottles desc
LIMIT 1
) as temp1

union 

SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2152
ORDER BY sale_bottles desc
LIMIT 1
) as temp2

union 

SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2190
ORDER BY sale_bottles desc
LIMIT 1
) as temp3

union 

SELECT STORE,sale_bottles
from(
SELECT STORE,sale_bottles
FROM iowa
WHERE STORE=2191
ORDER BY sale_bottles desc
LIMIT 1
) as temp4"""
}

### __(3 points) Q1.4: This problem has two steps, you only need to return the value from second step.__

__Compute the set of all liquors with the characters "tequila" (in lowercase) in its description (`im_desc`). The attribute `itemno` is the one that can serve as a unique identifier for a specific liquor.__

__Return the count of all distinct value of zipcode for all liquor stores that sold at least two of every tequila as defined in the previous sentence.__

(Optional: It is possible to rewrite the query in a form that can execute quite fast, but it will require you to create some temporary tables and/or experiment with alternate forms to express the problem. If you find the solution, feel free to submit it!)

In [25]:
%%sql
with temp1 as(
select store,zipcode,itemno,sale_bottles from iowa where itemno in (select distinct itemno from iowa where im_desc like '%tequila%')),
temp2 as(select store,zipcode,itemno,sale_bottles from temp1 T1 
where not exists((select distinct itemno from iowa where im_desc like '%tequila%') except (select itemno from temp1 T2 WHERE T2.store=T1.store))),
temp3 as(
select store, itemno,sum(sale_bottles) from temp2 group by store, itemno having sum(sale_bottles)>=2)
select count(distinct zipcode) from temp2 where store in (select distinct store from temp3)

1 rows affected.


count
0    115

In [34]:
# please replace None with the according count here
# submit query in string format
q4 = {
    'count': 115,
    'query': """with temp1 as(
select store,zipcode,itemno,sale_bottles from iowa where itemno in (select distinct itemno from iowa where im_desc like '%tequila%')),
temp2 as(select store,zipcode,itemno,sale_bottles from temp1 T1 
where not exists((select distinct itemno from iowa where im_desc like '%tequila%') except (select itemno from temp1 T2 WHERE T2.store=T1.store))),
temp3 as(
select store, itemno,sum(sale_bottles) from temp2 group by store, itemno having sum(sale_bottles)>=2)
select count(distinct zipcode) from temp2 where store in (select distinct store from temp3)"""
}

### __ (4 points) Q1.5:  __Compute the median itemno for the store you get in Q1.1

Let's take the store 4396 for example, if you run the query below, you would get the itemno and how many times each itemno showed.

In [35]:
%%sql
select itemno, count(1) as count
from iowa
where store = 4396
group by itemno
order by itemno

65 rows affected.


itemno  count
0     8828     12
1    11296      3
2    11776      4
3    11777      2
4    11788     14
5    12408      5
6    12888      1
7    15626      1
8    16518      4
9    16987      4
10   19066      2
11   24458      7
12   25606      2
13   25608      1
14   26824      2
15   26826      1
16   27102      1
17   27479      1
18   27544      1
19   28866      2
20   32237      1
21   34006      2
22   34433      1
23   35318      1
24   35914      7
25   35918      4
26   36307      3
27   36447      1
28   37938     16
29   37996      4
..     ...    ...
35   43127      1
36   43136      1
37   43302      1
38   43336      1
39   43337      3
40   43338      5
41   44350      1
42   45276      2
43   45278      3
44   50686      1
45   52316      4
46   57125      2
47   57144      1
48   58838      2
49   58888      1
50   64866      1
51   68036      1
52   69637      1
53   69656      1
54   71886      1
55   73050      2
56   75210      2
57   75214      1
58   76486      3
59   81206      1
60   82846      2
61   85526      1
62   86886      1
63   87936      1
64   89196      2

[65 rows x 2 columns]

Therefore, the median is computed as the median of the array \[8828, 8828, ... (totally 12 times of 8828), 11296, 11296, 11296, 11776, 11776, 11776, 11776, ..., 89196, 89196\].

__you are asked to write a query to compute the median of the store that you got from Q1.1. Hint: you might want to substitute '4396' with the result of Q1.1 and use that as a temporary table (with clause, subquery or view) first and then see how to compute the median.__

In [21]:
%%sql
-- Complete your query here
select itemno
from iowa
where store = 2190
order by itemno


7802 rows affected.


itemno
0       352
1       352
2       352
3       456
4       780
5       796
6       796
7       830
8       832
9       837
10      859
11     1014
12     1014
13     1014
14     1343
15     1799
16     1799
17     1799
18     1799
19     1931
20     1931
21     1931
22     2229
23     2249
24     2368
25     2368
26     2380
27     2380
28     2878
29     2878
..      ...
970   16517
971   16517
972   16517
973   16517
974   16517
975   16517
976   16517
977   16562
978   16562
979   16562
980   16562
981   16580
982   16676
983   16676
984   16676
985   16676
986   16676
987   16676
988   16676
989   16850
990   16850
991   16850
992   16850
993   16850
994   16906
995   16987
996   16987
997   17086
998   17086
999   17086

[1000 rows x 1 columns]

In [35]:
%%sql
-- Complete your query here

with temp1 as(
select itemno
from iowa
where store = 2190
order by itemno
OFFSET (select count(itemno)/2+count(itemno)%2-1 from iowa where store = 2190) ROWS
FETCH NEXT 2 ROWS only)

select avg(itemno)::int from temp1

1 rows affected.


avg
0  41078

In [37]:
# please replace None with the item number here
# submit query in string format
q5 = {
    'median': 41078,
    'query': """
    with temp1 as(
select itemno
from iowa
where store = 2190
order by itemno
OFFSET (select count(itemno)/2+count(itemno)%2-1 from iowa where store = 2190) ROWS
FETCH NEXT 2 ROWS only)
select avg(itemno)::int from temp1
    """
}

### __(3 points) Q1.6: There are some subtle difference between 'not exist' and 'not in'. Give a scenario that you want to use 'not exist' rather than 'not in' and vice versa (could be in either SQL or English as long as it makes sense). __

In [38]:
q6 = {
    # please write in the quotes
    'paragraph': """ NOT EXIST(subquery) is faster than NOT IN(subquery).
    If the subquery returned any rows, then when evaluating WHERE [NOT] EXISTS (subquery), the database engine could quit searching as soon as it had found just one row, whereas WHERE [NOT] IN (subquery) would always collect all the results from the sub-query, before further processing.
    Thus, if the subquery is large, I prefer using "not exists". 
    NOT IN(subquery) is sensitive to null. If there is null in the subquery, then NOT IN(subquery) will only return 'False'or'NULL'. 
    """
}

## Part I Submission

### Create your submission file¶

Run the following cell to create a results file for your homework

DO NOT MODIFY THE FOLLOWING CELL!!

In [39]:
import datetime
import json

script_path = '{0}/{1}/fs/Instabase%20Drive'.format(USER, REPO)

with ib.open('results'.format(script_path), "w") as f:
    result = dict(
        q1=q1,
        q2=q2,
        q3=q3,
        q4=q4,
        q5=q5,
        q6=q6,
        uni=UNI,
        user=USER
    )
    f.write(json.dumps(result))
    print "Result file created at: {0}".format(datetime.datetime.now())
  
    print
    print "Check your results: http://instabase.com/{0}/HW3/results".format(script_path)

Result file created at: 2019-04-02 02:57:51.198962

Check your results: http://instabase.com/zd2221/myrepo/fs/Instabase%20Drive/HW3/results


Finally, Submit your __HW3 folder including the results file__ at the following URL:
    
https://www.instabase.com/apps/file-submission/cmd/submit/4d4994ed-a75e-4240-9499-c098baee7c58


## Part II, Normalization

See Github page for more details. Submit __hard copy__ for Part 2 as instructed __before due__.

_**Remarks: Dataset sourced from https://data.iowa.gov/Economy/Iowa-Liquor-Sales/m3tr-qhgy. 1mil tuples extracted and preprocessed to remove double quotation marks and apostrophes._